### PR TITLE
human readable rule traces

### PR DESCRIPTION
**Pull request by Soph1514**
Added functionality to capture rule traces in a nicer format to a txt file in integration tests. 

from `conjure_oxide/tests/integration/xyz/input-generated-rule-trace-human.txt`
```rust
Initial expression: Sum([Sum([a, b]), c]), 
     Rule: normalise_associative_commutative ([("Base", 8400)]) 
Result: Sum([a, b, c]) 


Initial expression: (Sum([a, b, c]) = 4), 
     Rule: sum_eq_to_sumeq ([("Minion", 4400)]) 
Result: SumEq([a, b, c], 4) 


Initial expression: SumEq([a, b, c], 4), 
     Rule: sumeq_to_minion ([("Minion", 4400)]) 
Result: And([SumGeq([a, b, c], 4), SumLeq([a, b, c], 4)]) 


Initial expression: And([And([SumGeq([a, b, c], 4), SumLeq([a, b, c], 4)]), (a >= b)]), 
     Rule: normalise_associative_commutative ([("Base", 8400)]) 
Result: And([SumGeq([a, b, c], 4), SumLeq([a, b, c], 4), (a >= b)]) 


Initial expression: (a >= b), 
     Rule: geq_to_ineq ([("Minion", 4100)]) 
Result: Ineq(b, a, 0) 




```

### CODE REVIEW

File: solvers/chuffed/vendor
Comment (niklasdewally) [inline solvers/chuffed/vendor:1]: This needs removing (see teams)

**Review by niklasdewally — APPROVED** [review]
LGTM!

**Comment by Soph1514** [comment]
now
```rust
Sum([Sum([a, b]), c]),
   ~~> normalise_associative_commutative ([("Base", 8400)])
Sum([a, b, c])
--

(Sum([a, b, c]) = 4),
   ~~> sum_eq_to_sumeq ([("Minion", 4400)])
SumEq([a, b, c], 4)
--

SumEq([a, b, c], 4),
   ~~> sumeq_to_minion ([("Minion", 4400)])
And([SumGeq([a, b, c], 4), SumLeq([a, b, c], 4)])
```

**Comment by ozgurakgun** [comment]
Looks great. Are we going to make these part of the testing generated/expected style?

We should also make it so -v in the main executale produces this style of output, but that can come in a separate PR.

**Comment by Soph1514** [comment]
I do not think it makes much sense since the content is basically the same as in the json files so if there is a mismatch between the expected and generated we will know from that test. But If you think it is necessary, I can do it!

**Comment by ozgurakgun** [comment]
So how do you want to incorporate this otherwise?

One argument for having it saved in the repo is to make PR reviews easy.

**Comment by Soph1514** [comment]
I guess I could just have the expected file with the human readable rules in the repo for us to be able to see the rules but not necessarily compare it with the generated one using asserts?

**Comment by ozgurakgun** [comment]
once you have it in the repo, why not compare as well?

one way of reducing the redundancy would be to remove the json traces, see #392 where they just caused an issue actually, but since they might contain more info in the future (like metadata) I am tempted to keep both.

File: crates/conjure_core/src/rule_engine/rewrite.rs [hunk 1]
```diff
@@ -439,6 +439,15 @@ fn trace_rules(results: &[RuleResult], expression: Expression) {
         let rule = results[0].rule;
         let new_expression = results[0].reduction.new_expression.clone();
 
+        trace!(
+            target: "rule_engine_human",
+            "{}, \n   ~~> {} ({:?}) \n{} \n--\n",
```
Comment (ozgurakgun) [inline crates/conjure_core/src/rule_engine/rewrite.rs:444]: can you change this to \n\n-- please? it's a personal preference I know...

**Review by ozgurakgun — COMMENTED** [review]
one minor change only - thanks!

File: crates/conjure_core/src/rule_engine/rewrite.rs (hunk 1 above)
Comment (ozgurakgun) [reply crates/conjure_core/src/rule_engine/rewrite.rs]: ooops sorry. I didn't want to remove the \n after --

to be clear I wanted: 


```
            "{}, \n   ~~> {} ({:?}) \n{} \n\n--\n",
```

File: conjure_oxide/tests/integration/basic/div/04/input-expected-rule-trace-human.txt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+Not((a = SafeDiv(b, c))), 
+   ~~> negated_eq_to_neq ([("Base", 8800)]) 
+(a != SafeDiv(b, c)) 
+
+--
+
+(a != SafeDiv(b, c)), 
+   ~~> flatten_binop ([("Minion", 4400)]) 
+(a != __0)
```
Comment (ozgurakgun) [inline conjure_oxide/tests/integration/basic/div/04/input-expected-rule-trace-human.txt:33]: Not for this PR, but something to be captured for future PRs.

This rule is adding a new aux var to the symbol table. In general rules can have global side effects. If so, we should display information about the changes being made to the symbol table here. This is a good example.

@Soph1514

Comment (Soph1514) [reply conjure_oxide/tests/integration/basic/div/04/input-expected-rule-trace-human.txt:33]: Good idea! On my agenda for future pr's

Comment (ozgurakgun) [reply conjure_oxide/tests/integration/basic/div/04/input-expected-rule-trace-human.txt:33]: I made an issue and assigned it to you, not sure if you noticed?

**Pull request merged**